libs/platform/ebpf_ring_buffer.c

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "ebpf_epoch.h"
+#include "ebpf_ring_buffer.h"
+
+typedef struct _ebpf_ring_buffer
+{
+    ebpf_lock_t lock;
+    size_t length;
+    size_t consumer_offset;
+    size_t producer_offset;
+    uint8_t* shared_buffer;
+    ebpf_ring_descriptor_t* ring_descriptor;
+} ebpf_ring_buffer_t;
+
+inline static size_t
+_ring_get_producer_offset(_In_ const ebpf_ring_buffer_t* ring)
+{
+    return ring->producer_offset % ring->length;
+}
+
+inline static size_t
+_ring_get_consumer_offset(_In_ const ebpf_ring_buffer_t* ring)
+{
+    return ring->consumer_offset % ring->length;
+}
+
+inline static void
+_ring_advance_producer_offset(_Inout_ ebpf_ring_buffer_t* ring, size_t length)
+{
+    ring->producer_offset += length;
+}
+
+inline static void
+_ring_advance_consumer_offset(_Inout_ ebpf_ring_buffer_t* ring, size_t length)
+{
+    ring->consumer_offset += length;
+}
+
+inline static _Ret_notnull_ ebpf_ring_buffer_record_t*
+_ring_record_at_offset(_In_ const ebpf_ring_buffer_t* ring, size_t offset)
+{
+    return (ebpf_ring_buffer_record_t*)&ring->shared_buffer[offset % ring->length];
+}
+
+inline static _Ret_notnull_ ebpf_ring_buffer_record_t*
+_ring_next_consumer_record(_In_ const ebpf_ring_buffer_t* ring)
+{
+    return _ring_record_at_offset(ring, _ring_get_consumer_offset(ring));
+}
+
+inline static _Ret_maybenull_ ebpf_ring_buffer_record_t*
+_ring_buffer_acquire_record(_Inout_ ebpf_ring_buffer_t* ring, size_t requested_length)
+{
+    ebpf_ring_buffer_record_t* record = NULL;
+    requested_length += EBPF_OFFSET_OF(ebpf_ring_buffer_record_t, data);
```

Comment:
so if caller passes requested_length == ring_buf_capacity, this fails. But passing setting requested_length = ring_buf_capacity should be allowed.
Maybe consider taking a record_length , max_count parameter in the create call instead so that the headers can be accomodated?

Reply:
It makes things simpler to have the caller pass in the count of bytes of memory requests and have one function compute the extra space required for the header.